In [1]:
import pandas as pd
import glob
import os
import matplotlib.pyplot as plt
import seaborn as sns

# Cài đặt style chung cho biểu đồ đẹp hơn
sns.set_theme(style="whitegrid")
plt.rcParams['figure.figsize'] = (10, 5)

# Sửa lại đường dẫn: lùi ra ngoài thư mục Notebooks để thấy processed_chunks
path = '../processed_chunks/' 

all_files = glob.glob(os.path.join(path, "train_*.pkl"))

# Kiểm tra xem có tìm thấy file nào không
if not all_files:
    print("Không tìm thấy file nào! Hãy kiểm tra lại đường dẫn.")
else:
    print(f"Tìm thấy {len(all_files)} files. Đang đọc...")
    df_list = [pd.read_pickle(f) for f in all_files]
    full_df = pd.concat(df_list, ignore_index=True)
    print(f"Thành công! Tổng số dòng: {len(full_df)}")


Tìm thấy 18 files. Đang đọc...
Thành công! Tổng số dòng: 1708337


## 1. Data understanding

In [2]:
full_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1708337 entries, 0 to 1708336
Data columns (total 59 columns):
 #   Column                                        Dtype         
---  ------                                        -----         
 0   channelGrouping                               object        
 1   date                                          datetime64[ns]
 2   fullVisitorId                                 object        
 3   visitId                                       int64         
 4   visitNumber                                   int64         
 5   visitStartTime                                int64         
 6   device_browser                                object        
 7   device_operatingSystem                        object        
 8   device_isMobile                               bool          
 9   device_deviceCategory                         object        
 10  geoNetwork_continent                          object        
 11  geoNetwork_subContinent            

In [3]:
full_df.tail(10)

,channelGrouping,date,fullVisitorId,visitId,visitNumber,visitStartTime,device_browser,device_operatingSystem,device_isMobile,device_deviceCategory,...,Date_Is_month_start,Date_Is_quarter_end,Date_Is_quarter_start,Date_Is_year_end,Date_Is_year_start,visitStartTime_datetime,Date_Hour,transactionRevenue,transactionRevenue_dollar,target_log_revenue
1708327,Referral,2018-02-23,2078818331430817585,1519409528,4,1519409528,Chrome,Windows,False,desktop,...,0,0,0,0,0,2018-02-23 18:12:08,18,0.0,0.00,0.000000
1708328,Organic Search,2018-02-23,0674894690195844684,1519418239,1,1519418239,Firefox,Macintosh,False,desktop,...,0,0,0,0,0,2018-02-23 20:37:19,20,0.0,0.00,0.000000
1708329,Organic Search,2018-02-23,6960314182921269906,1519408746,5,1519408746,Chrome,Windows,False,desktop,...,0,0,0,0,0,2018-02-23 17:59:06,17,0.0,0.00,0.000000
1708330,Referral,2018-02-23,19627086332989220,1519452507,3,1519452507,Chrome,Macintosh,False,desktop,...,0,0,0,0,0,2018-02-24 06:08:27,6,0.0,0.00,0.000000
1708331,Referral,2018-02-23,9561952862629294086,1519422952,1,1519422952,Chrome,Linux,False,desktop,...,0,0,0,0,0,2018-02-23 21:55:52,21,95950000.0,95.95,18.379338
1708332,Organic Search,2018-02-23,3725326097124390672,1519396864,2,1519396864,Chrome,Windows,False,desktop,...,0,0,0,0,0,2018-02-23 14:41:04,14,0.0,0.00,0.000000
1708333,Referral,2018-02-23,2158552115222657286,1519434802,1,1519434802,Chrome,Macintosh,False,desktop,...,0,0,0,0,0,2018-02-24 01:13:22,1,79900000.0,79.90,18.196286
1708334,Referral,2018-02-23,8157991262626030554,1519416006,1,1519416006,Chrome,Android,True,mobile,...,0,0,0,0,0,2018-02-23 20:00:06,20,0.0,0.00,0.000000
1708335,Direct,2018-02-23,9963007161745063779,1519379680,1,1519379680,Chrome,Windows,False,desktop,...,0,0,0,0,0,2018-02-23 09:54:40,9,0.0,0.00,0.000000
1708336,Referral,2018-02-23,8111333210792218028,1519415541,1,1519415541,Chrome,Macintosh,False,desktop,...,0,0,0,0,0,2018-02-23 19:52:21,19,105570000.0,105.57,18.474885


In [4]:
full_df.isnull().sum()

channelGrouping                                       0
date                                                  0
fullVisitorId                                         0
visitId                                               0
visitNumber                                           0
visitStartTime                                        0
device_browser                                       12
device_operatingSystem                            11815
device_isMobile                                       0
device_deviceCategory                                 0
geoNetwork_continent                               2517
geoNetwork_subContinent                            2517
geoNetwork_country                                 2517
geoNetwork_region                                982733
geoNetwork_metro                                1319855
geoNetwork_city                                  998826
geoNetwork_networkDomain                         768845
totals_visits                                   

In [5]:
full_df.nunique().sort_values(ascending=False)

visitStartTime                                  1667423
visitStartTime_datetime                         1667423
visitId                                         1665802
fullVisitorId                                   1323730
trafficSource_adwordsClickInfo.gclId              59009
geoNetwork_networkDomain                          41980
totals_totalTransactionRevenue                     8507
target_log_revenue                                 7252
transactionRevenue                                 7252
totals_transactionRevenue                          7252
transactionRevenue_dollar                          7252
totals_timeOnSite                                  4775
trafficSource_keyword                              4546
trafficSource_referralPath                         3197
geoNetwork_city                                     954
date                                                638
geoNetwork_region                                   481
visitNumber                                     

## 📊 Data Dictionary — Google Merchandise Store

**Dataset:** Google Analytics session-level data
**Shape:** 1,708,337 rows × 59 columns · **Memory:** ~697 MB
**Target:** `target_log_revenue` — log1p(transaction revenue)

---

### 1. 🔑 Session Identification (6 columns)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `fullVisitorId` | object | 0 | Unique ID for each visitor |
| `visitId` | int64 | 0 | Session ID (unique per visitor) |
| `visitNumber` | int64 | 0 | Nth visit of this visitor (1 = first visit) |
| `visitStartTime` | int64 | 0 | UNIX timestamp of session start |
| `visitStartTime_datetime` | datetime64 | 0 | Human-readable session start time |
| `date` | datetime64 | 0 | Date of the session |

---

### 2. 📱 Device Information (4 columns)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `device_browser` | object | 12 | Browser used (Chrome, Safari, ...) |
| `device_operatingSystem` | object | 11,815 | OS of the device |
| `device_isMobile` | bool | 0 | True if session was from mobile |
| `device_deviceCategory` | object | 0 | Device type: desktop / mobile / tablet |

---

### 3. 🌍 Geographic Information (7 columns)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `geoNetwork_continent` | object | 2,517 | Continent of the visitor |
| `geoNetwork_subContinent` | object | 2,517 | Sub-continent |
| `geoNetwork_country` | object | 2,517 | Country of the visitor |
| `geoNetwork_region` | object | 982,733 | 🟡 Region/state — high nulls |
| `geoNetwork_metro` | object | 1,319,855 | 🔴 Metro area — very high nulls |
| `geoNetwork_city` | object | 998,826 | 🔴 City — very high nulls |
| `geoNetwork_networkDomain` | object | 768,845 | 🟡 ISP/network domain |

---

### 4. 📈 Session Totals (11 columns)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `totals_visits` | int64 | 0 | Always 1 (one row = one session) |
| `totals_hits` | int64 | 0 | Number of hits in session |
| `totals_pageviews` | int64 | 0 | Pages viewed in session |
| `totals_bounces` | int64 | 0 | 1 if session bounced, else 0 |
| `totals_newVisits` | int64 | 0 | 1 if first-ever visit, else 0 |
| `totals_sessionQualityDim` | int64 | 0 | GA session quality score (1–100) |
| `totals_timeOnSite` | int64 | 0 | Time on site in seconds |
| `totals_transactions` | int64 | 0 | Number of transactions in session |
| `totals_transactionRevenue` | float64 | 0 | Revenue × 10⁶ (raw GA format) |
| `totals_totalTransactionRevenue` | float64 | 0 | Total revenue × 10⁶ |
| `transactionRevenue` | float64 | 0 | Revenue cleaned (= totals / 10⁶) |

---

### 5. 🚦 Traffic Source (13 columns)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `channelGrouping` | object | 0 | Channel: Organic, Direct, Referral, ... |
| `trafficSource_source` | object | 70 | Traffic source (google, direct, ...) |
| `trafficSource_medium` | object | 566,091 | 🟡 Medium: organic, cpc, referral, ... |
| `trafficSource_campaign` | object | 1,604,526 | 🔴 Campaign name — mostly null |
| `trafficSource_keyword` | object | 0 | Search keyword |
| `trafficSource_referralPath` | object | 0 | Referral URL path |
| `trafficSource_isTrueDirect` | bool | 0 | True if confirmed direct traffic |
| `trafficSource_adContent` | object | 0 | Ad content label |
| `trafficSource_adwordsClickInfo.page` | int64 | 0 | AdWords page number |
| `trafficSource_adwordsClickInfo.slot` | object | 0 | Ad slot position |
| `trafficSource_adwordsClickInfo.gclId` | object | 0 | Google Click ID |
| `trafficSource_adwordsClickInfo.adNetworkType` | object | 0 | Ad network type |
| `trafficSource_adwordsClickInfo.isVideoAd` | bool | 0 | True if video ad click |

---

### 6. 🗓️ Date Features (13 columns — engineered)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `Date_Year` | int32 | 0 | Year |
| `Date_Month` | int32 | 0 | Month (1–12) |
| `Date_Day` | int32 | 0 | Day of month (1–31) |
| `Date_Dayofweek` | int32 | 0 | Day of week (0=Mon, 6=Sun) |
| `Date_Dayofyear` | int32 | 0 | Day of year (1–365) |
| `Date_Week` | Int64 | 0 | ISO week number |
| `Date_Hour` | int32 | 0 | Hour of visit (0–23) |
| `Date_Is_month_end` | int64 | 0 | 1 if last day of month |
| `Date_Is_month_start` | int64 | 0 | 1 if first day of month |
| `Date_Is_quarter_end` | int64 | 0 | 1 if last day of quarter |
| `Date_Is_quarter_start` | int64 | 0 | 1 if first day of quarter |
| `Date_Is_year_end` | int64 | 0 | 1 if last day of year |
| `Date_Is_year_start` | int64 | 0 | 1 if first day of year |

---

### 7. 🎯 Target & Custom Dimensions (5 columns)

| Column | Type | Nulls | Description |
|---|---|---|---|
| `transactionRevenue_dollar` | float64 | 0 | Revenue in USD (cleaned) |
| `target_log_revenue` | float64 | 0 | **Model target** — log1p(revenue) |
| `customDimensions_count` | int64 | 0 | Number of custom dimensions |
| `customDimensions_index` | int64 | 0 | Custom dimension index |
| `customDimensions_value` | object | 0 | Custom dimension value |

---

### ⚠️ Missing Value — Cách xử lý

| Severity | Columns | Cách xử lý |
|---|---|---|
| 🔴 Rất cao (>50%) | `trafficSource_campaign` | Fill `"no_campaign"` — null = không chạy campaign |
| 🔴🟡 Cao | `geoNetwork_metro`, `geoNetwork_city`, `geoNetwork_region` | Fill `"unknown"` |
| 🟡 Cao (~33%) | `trafficSource_medium` | Fill `"none"` — chuẩn GA: null = direct/none |
| 🟡 Cao (~45%) | `geoNetwork_networkDomain` | Không dùng trong model — bỏ qua |
| 🟢 Thấp (<1%) | `device_operatingSystem`, `device_browser`, `trafficSource_source`, `geoNetwork_country`, `geoNetwork_subContinent`, `geoNetwork_continent` | Fill `"unknown"` |

> **Kết quả:** Sau xử lý tất cả cột đều có null = 0.

Không nên impute vì bịa dữ liệu,không nên drop vì mất data để predict trong khi các cột dữ liệu khác vẫn còn. Ngoài ra, null còn có thể hiểu là users ẩn thông tin or dùng browers lạ. Tỷ lệ null cao ở city, country... là do không track được

In [6]:
# Danh sách các cột bạn đã gửi
cols_to_check = [
    'fullVisitorId', 'trafficSource_adwordsClickInfo.gclId', 'geoNetwork_networkDomain',
    'trafficSource_keyword', 'trafficSource_referralPath', 'geoNetwork_city',
    'geoNetwork_region', 'trafficSource_source', 'geoNetwork_country',
    'device_browser', 'geoNetwork_metro', 'trafficSource_adContent',
    'trafficSource_campaign', 'device_operatingSystem', 'geoNetwork_subContinent',
    'channelGrouping', 'customDimensions_value', 'geoNetwork_continent',
    'trafficSource_medium', 'trafficSource_adwordsClickInfo.slot',
    'trafficSource_adwordsClickInfo.adNetworkType', 'device_deviceCategory'
]

# Kiểm tra các cột có tồn tại trong DataFrame không trước khi check
existing_cols = [c for c in cols_to_check if c in full_df.columns]

# Tính toán số lượng Null và tỷ lệ %
null_info = pd.DataFrame({
    'Số lượng Null': full_df[existing_cols].isnull().sum(),
    'Tỷ lệ %': (full_df[existing_cols].isnull().sum() / len(full_df)) * 100
})

print("--- Thống kê giá trị thiếu (Null/NA) ---")
print(null_info.sort_values(by='Số lượng Null', ascending=False))


--- Thống kê giá trị thiếu (Null/NA) ---
                                              Số lượng Null    Tỷ lệ %
trafficSource_campaign                              1604526  93.923272
geoNetwork_metro                                    1319855  77.259639
geoNetwork_city                                      998826  58.467738
geoNetwork_region                                    982733  57.525711
geoNetwork_networkDomain                             768845  45.005464
trafficSource_medium                                 566091  33.136963
device_operatingSystem                                11815   0.691608
geoNetwork_country                                     2517   0.147336
geoNetwork_subContinent                                2517   0.147336
geoNetwork_continent                                   2517   0.147336
trafficSource_source                                     70   0.004098
device_browser                                           12   0.000702
fullVisitorId                       

## 2. Data cleaning - xử lý null, outliers, noise cơ bản

Xử lý null/NA

In [7]:
# BƯỚC 1 — Null cực cao: điền giá trị có nghĩa
full_df['trafficSource_campaign'] = full_df['trafficSource_campaign'].fillna('no_campaign')
full_df['geoNetwork_metro']       = full_df['geoNetwork_metro'].fillna('unknown')
full_df['geoNetwork_city']        = full_df['geoNetwork_city'].fillna('unknown')
full_df['geoNetwork_region']      = full_df['geoNetwork_region'].fillna('unknown')

# BƯỚC 2 — trafficSource_medium: null = direct theo chuẩn GA
full_df['trafficSource_medium'] = full_df['trafficSource_medium'].fillna('none')

# BƯỚC 3 — Null nhỏ < 1%: điền unknown
small_null_cols = [
    'device_operatingSystem',
    'geoNetwork_country',
    'geoNetwork_subContinent',
    'geoNetwork_continent',
    'trafficSource_source',
    'device_browser',
]
for col in small_null_cols:
    full_df[col] = full_df[col].fillna('unknown')

# BƯỚC 4 — Kiểm tra lại
print(full_df[existing_cols].isnull().sum().sort_values(ascending=False))

geoNetwork_networkDomain                        768845
fullVisitorId                                        0
trafficSource_adwordsClickInfo.gclId                 0
trafficSource_keyword                                0
trafficSource_referralPath                           0
geoNetwork_city                                      0
geoNetwork_region                                    0
trafficSource_source                                 0
geoNetwork_country                                   0
device_browser                                       0
geoNetwork_metro                                     0
trafficSource_adContent                              0
trafficSource_campaign                               0
device_operatingSystem                               0
geoNetwork_subContinent                              0
channelGrouping                                      0
customDimensions_value                               0
geoNetwork_continent                                 0
trafficSou

In [23]:
rare = full_df['device_browser'].value_counts()
rare = rare[rare <5].index

In [24]:
full_df[full_df['device_browser'].isin(rare)]['device_browser']

4099                                                     ADM
25244                                   IE with Chrome Frame
32801                                            ThumbSniper
144580     ;__CT_JOB_ID__:65da7e5f-0f05-4b5d-8d31-1f4d470...
161115     ;__CT_JOB_ID__:a80e8e16-6e98-455b-885a-a4dd40f...
                                 ...                        
1688957    ;__CT_JOB_ID__:85da5736-a78e-45a9-837e-f5a53e5...
1688965    ;__CT_JOB_ID__:a7ed0808-e70c-4b19-b1a3-1018bbb...
1690321    ;__CT_JOB_ID__:2e0eca60-83ab-482d-bb81-343d113...
1696476                           ecgiwapzltrkujdhmqsbxfonvy
1705128                                       LYF_LS_4002_11
Name: device_browser, Length: 116, dtype: object

##### noise, outliers: loại bỏ các session không phải người dùng thật

loại 1: internal traffic (nhân viên test): network_domain chứa google or city = Mountain View

In [28]:
# 1. Internal traffic (nhân viên test). Dấu hiệu: domain chứa google hoặc city = Mountain View -> test hệ thống, không mua thật 
internal_mask = (
    (full_df['geoNetwork_networkDomain'].astype(str).str.contains('google', na=False)) | 
    (full_df['geoNetwork_city'] == 'Mountain View')
)
print(f'Internal sessions: {internal_mask.sum():,}')
# 2. Bot/crawler: Các chương trình tự động crawl website, không phải người thật. Chúng để lại dấu hiệu bất thường như hits cao nhưng timeOnSite = 0
# pageview cao + bounce = 1 (vào rồi thoát ngay nhưng hits nhiều), browser lạ như ";__CT_JOB_ID..." -> bot không bao giờ mua hàng
bot_mask = (
    ((full_df['totals_hits'] > 500) & (full_df['totals_timeOnSite'] == 0)) |
    ((full_df['totals_pageviews'] > 500) & (full_df['totals_bounces'] == 1)) |
    (full_df['device_browser'].str.contains('CT_JOB_ID|ecgiwap', na=False, regex=True)))
print(f"Bot sessions: {bot_mask.sum():,}")

# 3. Tracking error: users chỉ xem 1 trang duy nhất rồi thoát (không thể vừa bounce vừa mua hàng vì cần min 3-4 pages: page sản phẩm -> giỏ hàng -> thanh toán -> xác nhận)
# bounce = 1 + revenue > 0 = lỗi tracking, không phải giao dịch thật
tracking_error_mask = ((full_df['totals_bounces'] == 1) & (full_df['transactionRevenue'] > 0))
print(f"Tracking errors: {tracking_error_mask.sum():,}")
print(f"Revenue từ nhóm này: ${full_df[tracking_error_mask]['transactionRevenue'].sum():,.2f}")


# 4. Fake Whale: pageviews = 1, timeonsite = 0, revenue lớn (chỉ xem 1 trang, không dành 1 giây nào trên size nhuwngmua hàng giá trị lớn)
fake_whale_mask = ((full_df['totals_pageviews'] <= 1) & (full_df['totals_timeOnSite']==0) & (full_df['transactionRevenue'] >0))
print(f"Fake whales: {fake_whale_mask.sum():,}")
print(f"Revenue trung bình: ${full_df[fake_whale_mask]['transactionRevenue'].mean():,.2f}")
print(f"Revenue max: ${full_df[fake_whale_mask]['transactionRevenue'].max():,.2f}")
print(f"Revenue tổng: ${full_df[fake_whale_mask]['transactionRevenue'].sum():,.2f}")

Internal sessions: 84,598
Bot sessions: 21
Tracking errors: 0
Revenue từ nhóm này: $0.00
Fake whales: 0
Revenue trung bình: $nan
Revenue max: $nan
Revenue tổng: $0.00


In [29]:
print(full_df[['totals_bounces', 'totals_timeOnSite', 
               'totals_pageviews', 'transactionRevenue']].describe())

print("\nNull counts:")
print(full_df[['totals_bounces', 'totals_timeOnSite']].isnull().sum())

# Kiểm tra có session bounce + revenue không
print("\nBounce = 1 và revenue > 0:")
print(((full_df['totals_bounces'] == 1) & 
       (full_df['transactionRevenue'] > 0)).sum())

# Kiểm tra pageviews = 1, time = 0
print("\nPageviews <= 1 và timeOnSite = 0:")
print(((full_df['totals_pageviews'] <= 1) & 
       (full_df['totals_timeOnSite'] == 0)).sum())

       totals_bounces  totals_timeOnSite  totals_pageviews  transactionRevenue
count    1.708337e+06       1.708337e+06      1.708337e+06        1.708337e+06
mean     5.101909e-01       1.238619e+02      3.695685e+00        1.355906e+06
std      4.998963e-01       3.544553e+02      6.472932e+00        4.522809e+07
min      0.000000e+00       0.000000e+00      0.000000e+00        0.000000e+00
25%      0.000000e+00       0.000000e+00      1.000000e+00        0.000000e+00
50%      1.000000e+00       0.000000e+00      1.000000e+00        0.000000e+00
75%      1.000000e+00       7.600000e+01      4.000000e+00        0.000000e+00
max      1.000000e+00       1.901700e+04      5.000000e+02        2.312950e+10

Null counts:
totals_bounces       0
totals_timeOnSite    0
dtype: int64

Bounce = 1 và revenue > 0:
0

Pageviews <= 1 và timeOnSite = 0:
872460


In [30]:
# Thử ngưỡng thấp hơn
for threshold in [500, 200, 100, 50]:
    n = ((full_df['totals_hits'] > threshold) & 
         (full_df['totals_timeOnSite'] == 0)).sum()
    print(f"hits > {threshold} & timeOnSite = 0: {n:,} sessions")

hits > 500 & timeOnSite = 0: 0 sessions
hits > 200 & timeOnSite = 0: 0 sessions
hits > 100 & timeOnSite = 0: 0 sessions
hits > 50 & timeOnSite = 0: 0 sessions


In [31]:
# Phân tích nhóm pageviews <= 1 và timeOnSite = 0
group = full_df[
    (full_df['totals_pageviews'] <= 1) & 
    (full_df['totals_timeOnSite'] == 0)
]

print(f"Tổng sessions   : {len(group):,} ({len(group)/len(full_df)*100:.1f}%)")
print(f"Có revenue > 0  : {(group['transactionRevenue'] > 0).sum():,}")
print(f"Tổng revenue    : ${group['transactionRevenue'].sum():,.2f}")
print(f"\nPhân phối bounces:")
print(group['totals_bounces'].value_counts())
print(f"\nPhân phối pageviews:")
print(group['totals_pageviews'].value_counts())

Tổng sessions   : 872,460 (51.1%)
Có revenue > 0  : 0
Tổng revenue    : $0.00

Phân phối bounces:
totals_bounces
1    871573
0       887
Name: count, dtype: int64

Phân phối pageviews:
totals_pageviews
1    872290
0       170
Name: count, dtype: int64


In [8]:
full_df['trafficSource_adContent'] = full_df['trafficSource_adContent'].replace('NA', 'no_ad')

In [ ]:
source_col = full_df['trafficSource_source'].str.lower()

# Danh sách các luật Regex
conditions = [
    source_col.str.contains('google', na=False),
    source_col.str.contains('youtube|yt\.be', na=False),
    source_col.str.contains('yahoo', na=False),
    source_col.str.contains('facebook|fb\.com|m\.facebook', na=False),
    source_col.str.contains('bing', na=False),
    source_col.str.contains('reddit', na=False),
    source_col.str.contains('pinterest', na=False),
    source_col.str.contains('baidu', na=False),
    source_col.isin(['(direct)', 'direct'])
]

# Tên tương ứng khi khớp với luật Regex ở trên
choices = [
    'google', 
    'youtube', 
    'yahoo', 
    'facebook', 
    'bing', 
    'reddit', 
    'pinterest', 
    'baidu',
    'direct'
]

# Áp dụng thay thế, những thằng nào không khớp thì giữ nguyên tên cũ (default=source_col)
full_df['trafficSource_source'] = np.select(conditions, choices, default=source_col)

# Kiểm tra lại kết quả
print("Top 10 Source sau khi chuẩn hóa:")
print(full_df['trafficSource_source'].value_counts().head(10))


GIẢM CHIỀU DỮ LIỆU

In [18]:
cols = ['geoNetwork_city',                                   
    'geoNetwork_region',                                  
    'trafficSource_source',                              
    'geoNetwork_country',                                  
    'device_browser',                                     
    'geoNetwork_metro',                                  
    'trafficSource_adContent',                           
    'trafficSource_campaign',                              
    'device_operatingSystem',                            
    'geoNetwork_subContinent'] 

# for col in cols:
#     print('========================================')
#     print('Value counts:\n')
#     print(full_df[col].value_counts())

In [10]:
def find_top_n_for_coverage(df, col, coverage = 0.9):
    'Tìm n nhỏ nhất để top n cover 90% tổng records'
    vc = df[col].value_counts(dropna=False, normalize=True).cumsum()
    n = (vc < coverage).sum()+1
    print(f"{col}: Top {n} cover 90% traffic")
    return n

for col in cols:
    find_top_n_for_coverage(full_df, col)

geoNetwork_city: Top 52 cover 90% traffic
geoNetwork_region: Top 36 cover 90% traffic
trafficSource_source: Top 3 cover 90% traffic
geoNetwork_country: Top 36 cover 90% traffic
device_browser: Top 3 cover 90% traffic
geoNetwork_metro: Top 3 cover 90% traffic
trafficSource_adContent: Top 1 cover 90% traffic
trafficSource_campaign: Top 1 cover 90% traffic
device_operatingSystem: Top 4 cover 90% traffic
geoNetwork_subContinent: Top 9 cover 90% traffic


In [11]:
# Số lượng tối thiểu để kiểm tra
check_threshold = 1000

cols_to_process = [
    'geoNetwork_city', 'geoNetwork_region', 'trafficSource_source',
    'geoNetwork_country', 'device_browser', 'geoNetwork_metro',
    'trafficSource_adContent', 'trafficSource_campaign',
    'device_operatingSystem', 'geoNetwork_subContinent',
]

print(f"--- Kiểm tra số lượng Category có > {check_threshold} records ---")
for col in cols_to_process:
    # Đếm tần suất của từng category trong cột
    counts = full_df[col].value_counts()
    
    # Lọc ra các category có tần suất > 100
    categories_over_threshold = counts[counts > check_threshold]
    
    # Đếm số lượng category thỏa mãn
    num_over = len(categories_over_threshold)
    total_categories = len(counts)
    
    print(f"Cột [{col:<25}]: {num_over:>4} / {total_categories:>4} categories (có > {check_threshold} records)")


--- Kiểm tra số lượng Category có > 1000 records ---
Cột [geoNetwork_city          ]:  102 /  955 categories (có > 1000 records)
Cột [geoNetwork_region        ]:   91 /  482 categories (có > 1000 records)
Cột [trafficSource_source     ]:   23 /  345 categories (có > 1000 records)
Cột [geoNetwork_country       ]:   81 /  228 categories (có > 1000 records)
Cột [device_browser           ]:   14 /  129 categories (có > 1000 records)
Cột [geoNetwork_metro         ]:   25 /  122 categories (có > 1000 records)
Cột [trafficSource_adContent  ]:    7 /   77 categories (có > 1000 records)
Cột [trafficSource_campaign   ]:    8 /   33 categories (có > 1000 records)
Cột [device_operatingSystem   ]:    8 /   24 categories (có > 1000 records)
Cột [geoNetwork_subContinent  ]:   19 /   23 categories (có > 1000 records)


In [12]:
# Dùng ngưỡng linh hoạt theo từng cột
thresholds = {
    # Geo chi tiết — cần ngưỡng cao hơn vì quá nhiều unique
    'geoNetwork_city':    5000,   # ~20-30 cities lớn
    'geoNetwork_region':  5000,   # ~30 regions
    'geoNetwork_metro':   1000,   # ~25 metros

    # Geo cấp cao — ngưỡng thấp hơn vì ít unique hơn
    'geoNetwork_country': 5000,   # ~81 countries — hoặc tăng lên 5000
    'geoNetwork_subContinent': 1000,  # 19/23, giữ gần hết

    # Traffic — ngưỡng 1000 là đủ
    'trafficSource_source':    1000,
    'trafficSource_adContent': 1000,
    'trafficSource_campaign':  1000,

    # Device — ngưỡng 1000 là đủ
    'device_browser':          1000,
    'device_operatingSystem':  1000,
}

print("--- Kiểm tra số lượng Category dựa trên ngưỡng tùy chỉnh ---")

# Lặp qua từng Cột và Ngưỡng tương ứng trong từ điển thresholds
for col, thresh in thresholds.items():
    # Đếm tần suất của từng category trong cột
    counts = full_df[col].value_counts()
    
    # Lọc ra các category có tần suất lớn hơn ngưỡng riêng (thresh) của cột đó
    categories_over_threshold = counts[counts > thresh]
    
    # Đếm số lượng category thỏa mãn
    num_over = len(categories_over_threshold)
    total_categories = len(counts)
    
    print(f"Cột [{col:<25}]: {num_over:>4} / {total_categories:>4} categories (có > {thresh:>4} records)")


--- Kiểm tra số lượng Category dựa trên ngưỡng tùy chỉnh ---
Cột [geoNetwork_city          ]:   40 /  955 categories (có > 5000 records)
Cột [geoNetwork_region        ]:   32 /  482 categories (có > 5000 records)
Cột [geoNetwork_metro         ]:   25 /  122 categories (có > 1000 records)
Cột [geoNetwork_country       ]:   46 /  228 categories (có > 5000 records)
Cột [geoNetwork_subContinent  ]:   19 /   23 categories (có > 1000 records)
Cột [trafficSource_source     ]:   23 /  345 categories (có > 1000 records)
Cột [trafficSource_adContent  ]:    7 /   77 categories (có > 1000 records)
Cột [trafficSource_campaign   ]:    8 /   33 categories (có > 1000 records)
Cột [device_browser           ]:   14 /  129 categories (có > 1000 records)
Cột [device_operatingSystem   ]:    8 /   24 categories (có > 1000 records)


In [13]:
# 1. Drop cột không cần thiết
cols_to_drop = [
       'trafficSource_adwordsClickInfo.gclId',
    'geoNetwork_networkDomain',
    'trafficSource_keyword',
    'trafficSource_referralPath'
    ]

full_df.drop(columns=cols_to_drop, inplace=True)

# 2. Top-n grouping để giảm cardinality
configs = {
    'geoNetwork_city': 40, 'geoNetwork_region': 30,
    'geoNetwork_metro': 25, 'trafficSource_source': 20,
    'geoNetwork_country': 40, 'device_browser': 14,
    'trafficSource_adContent': 7, 'trafficSource_campaign': 8,
    'geoNetwork_subContinent': 19, 'device_operatingSystem': 8
}

for col, n in configs.items():
    top = full_df[col].value_counts(dropna=False).nlargest(n).index
    full_df[col] = full_df[col].where(full_df[col].isin(top), other='Other')
